### ISSUE TITLE
.npmrc not properly applied when running Yarn via script

### ISSUE THREAD

**Issue by matoilic** [yarnpkg/yarn#4475]
When running Yarn through a script in `package.json`, it seems to ignore properties from the `.npmrc` file in the project folder. 

**Example**

_package.json_
```
{
  "name": "my-project",
  "license": "UNLICENSED",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
  },
  "devDependencies": {
  },
  "scripts": {
    "showconfig": "yarn config list"
  }
}
```

_.npmrc_
```
registry=https://example.com/repository/npm/
always-auth=true
_auth=xxxx
```

When running `yarn config list` I get back
```
yarn config v1.0.2
info yarn config
{ 'version-tag-prefix': 'v',
  'version-git-tag': true,
  'version-git-sign': false,
  'version-git-message': 'v%s',
  'init-version': '1.0.0',
  'init-license': 'MIT',
  'save-prefix': '^',
  'ignore-scripts': false,
  'ignore-optional': false,
  registry: 'https://registry.yarnpkg.com',
  'strict-ssl': true,
  'user-agent': 'yarn/1.0.2 npm/? node/v8.5.0 darwin x64',
  lastUpdateCheck: 1505395389597 }
info npm config
{ registry: 'https://example.com/repository/npm/',
  'always-auth': true,
  _auth: 'xxxx' }
```
But when running `yarn config list` via the `showconfig` script – `yarn run showconfig` – I get back

```
yarn run v1.0.2
$ yarn config list
{ 'version-tag-prefix': 'v',
  'version-git-tag': true,
  'version-git-sign': false,
  'version-git-message': 'v%s',
  'init-version': '1.0.0',
  'init-license': 'MIT',
  'save-prefix': '^',
  'ignore-scripts': false,
  'ignore-optional': false,
  registry: 'https://registry.yarnpkg.com',
  'strict-ssl': true,
  'user-agent': 'yarn/1.0.2 npm/? node/v8.5.0 darwin x64',
  lastUpdateCheck: 1505395389597,
  silent: '1' }
{ registry: 'https://registry.yarnpkg.com',
  'always-auth': true,
  _auth: 'xxx',
  silent: '1',
  auth: 'xxxx',
  'version-git-tag': true,
  'init-license': 'MIT',
  argv: '{"remain":[],"cooked":["setup"],"original":["setup"]}',
  'save-prefix': '^',
  'strict-ssl': true,
  'version-git-message': 'v%s',
  'version-git-sign': '',
  'ignore-scripts': '',
  'user-agent': 'yarn/1.0.2 npm/? node/v8.5.0 darwin x64',
  'init-version': '1.0.0',
  'ignore-optional': '',
  'version-tag-prefix': 'v' }
```

It seems like it loads the global config once more in the second case and merges it into the NPM config. The same happens when running other Yarn commands, e.g. `install`. Shouldn't the config be the same in both cases?

**Node version:** v8.5.0
**Yarn version:** 1.0.2
**OS**: macOS 10.12.6 (also tested on CentOS 7)

**Comment by jgoz** [comment]
I believe this is happening because the outer `yarn` adds `npm_*` environment variables for its current configuration, which are picked up by the inner `yarn` and merged into the config.

Due to the way `.npmrc` configuration is merged (first-config-wins), and due to the fact that environment variables are read first, this means whatever is defined in the `npm_*` variables will take precedence over the `.npmrc` files in the inner instance.

**Comment by matoilic** [comment]
OK, I see the point. Was that changed with Yarn 1.x? Because our build worked fine with 0.27.5.

**Comment by jgoz** [comment]
I think it was due to changes in Yarn 1.x, but I'm not sure what the exact change was. Same observation in our environment - did not see this behaviour in 0.27.5.

**Comment by roastlechon** [comment]
a couple of my coworkers have this issue. any thoughts on how to resolve?

**Comment by matoilic** [comment]
The solution for us was to just not run Yarn via a script but directly. Luckily, our setup was flexible enough to allow such a change without any big refactorings.

**Comment by jgoz** [comment]
We are running `yarn` via a Node child process, so the solution we used was to strip `npm_` variables from the current `env` and pass that to the child process.

**Comment by sebastianleezy** [comment]
I am encountering this issue and for those wondering how to unset `npm_`. This post from https://github.com/npm/npm/issues/17674 helped me.

TLDR:
```
unset $(env | awk -F= '$1 ~ /^npm_/ {print $1}')
```

**Comment by paul-soporan** [comment]
Closing as in v2 we completely stopped reading `.npmrc` files.

https://yarnpkg.com/getting-started/migration
